src/workerd/jsg/async-context.c++

```diff
@@ -0,0 +1,297 @@
+}
+
+v8::Local<v8::Function> AsyncContextFrame::attachContext(
+    Lock& js,
+    v8::Local<v8::Function> fn,
+    kj::Maybe<AsyncContextFrame&> maybeFrame,
+    kj::Maybe<v8::Local<v8::Value>> thisArg) {
+  auto isolate = js.v8Isolate;
+  auto context = isolate->GetCurrentContext();
+  auto handle = js.getPrivateSymbolFor("asyncResource"_kj);
+
+  // Let's make sure the given function has not already been wrapped. If it has,
+  // we'll explicitly throw an error since wrapping a function more than once is
+  // most likely a bug.
+  JSG_REQUIRE(!check(fn->HasPrivate(context, handle)), TypeError,
+      "This function has already been associated with an async context.");
+
+  // Because we are working directly with JS functions here and not jsg::Function,
+  // we do not have the option of using either an internal field or lambda capture.
+  // Instead, we create an opaque wrapper holding a ref to the current frame and set
+  // it as a private field on the function.
+  auto frame = kj::addRef(AsyncContextFrame::current(js));
+  KJ_ASSERT(check(fn->SetPrivate(context, handle, wrapOpaque(context, kj::mv(frame)))));
+  KJ_IF_MAYBE(arg, thisArg) {
+    auto thisArgHandle = js.getPrivateSymbolFor("thisArg"_kj);
+    KJ_ASSERT(check(fn->SetPrivate(context, thisArgHandle, *arg)));
+  }
+
+  return jsg::check(v8::Function::New(context,
+      [](const v8::FunctionCallbackInfo<v8::Value>& args) {
+    auto isolate = args.GetIsolate();
+    auto context = isolate->GetCurrentContext();
+    auto& js = Lock::from(isolate);
+    auto fn = args.Data().As<v8::Function>();
+    auto handle = js.getPrivateSymbolFor("asyncResource"_kj);
+    auto thisArgHandle = js.getPrivateSymbolFor("thisArg"_kj);
+
+    // We do not use the normal unwrapOpaque here since that would consume the wrapped
+    // value, and we need to be able to unwrap multiple times.
+    auto& frame = KJ_ASSERT_NONNULL(tryGetContextFrame(isolate,
+        check(fn->GetPrivate(context, handle))));
+
+    v8::Local<v8::Value> thisArg = context->Global();
+    if (fn->HasPrivate(context, thisArgHandle).FromJust()) {
+      thisArg = check(fn->GetPrivate(context, thisArgHandle));
+    }
+
+    AsyncContextFrame::Scope scope(js, frame);
+
+    kj::Vector<v8::Local<v8::Value>> argv(args.Length());
+    for (int n = 0; n < args.Length(); n++) {
+      argv.add(args[n]);
+    }
+
+    v8::Local<v8::Value> result;
+    if (fn->Call(context, thisArg, args.Length(), argv.begin()).ToLocal(&result)) {
+      args.GetReturnValue().Set(result);
+    }
+  }, fn));
+}
+
+v8::Local<v8::Promise> AsyncContextFrame::attachContext(
+    Lock& js,
+    v8::Local<v8::Promise> promise,
+    kj::Maybe<AsyncContextFrame&> maybeFrame) {
+  auto handle = js.getPrivateSymbolFor("asyncResource"_kj);
+  auto context = js.v8Isolate->GetCurrentContext();
+  // If the promise has already been wrapped, do nothing else and just return the promise.
```

Comment:
When would this happen? The kResolve case already checks if a context is attached before trying to attach one, so that leaves only the kInit call site. In that case, there should never already be a context attached, right?

If we don't expect we'll ever hit this then I'd say we should skip the check.

Reply:
Yeah it's purely defensive. I'll convert the check into a `KJ_DASSERT`